vdr/didsubject/did_document.go

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2024 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package didsubject
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/google/uuid"
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/jsonld"
+	"gorm.io/gorm"
+	"gorm.io/gorm/schema"
+	"time"
+)
+
+// DIDDocument is the gorm representation of the did_document_version table
+type DIDDocument struct {
+	ID        string `gorm:"primaryKey"`
+	DidID     string `gorm:"column:did"`
+	DID       DID    `gorm:"foreignKey:DidID;references:ID"`
+	CreatedAt int64  `gorm:"autoCreateTime:false"`
+	// UpdatedAt is the (unix) timestamp when the document was updated
+	// Also used to purge DID document changes that haven't been committed within a certain time frame
+	UpdatedAt           int64 `gorm:"autoUpdateTime:false"`
+	Version             int
+	VerificationMethods []VerificationMethod `gorm:"foreignKey:DIDDocumentID;references:ID"`
+	Services            []SqlService         `gorm:"foreignKey:DIDDocumentID;references:ID"`
```

Comment:
I think the values for `foreignKey` and `references` are reversed

Reply:
and still it works:
```
time="2024-07-10T14:30:18+02:00" level=trace msg="Query (took 353.291µs): SELECT * FROM `did_document_version` WHERE did = \"did:web:example.com:iam:bob\" ORDER BY version desc,`did_document_version`.`id` LIMIT 1" module=Storage
time="2024-07-10T14:30:18+02:00" level=trace msg="Query (took 334.792µs): INSERT INTO `did` (`id`,`subject`) VALUES (\"did:web:example.com:iam:bob\",\"bob\") ON CONFLICT DO NOTHING" module=Storage
time="2024-07-10T14:30:18+02:00" level=trace msg="Query (took 48.25µs): INSERT INTO `did_verificationmethod` (`id`,`did_document_id`,`key_types`,`data`) VALUES (\"#1\",\"125fedb4-ee25-4933-a28b-55c01c2e435a\",31,\"{}\") ON CONFLICT (`id`) DO UPDATE SET `did_document_id`=`excluded`.`did_document_id`" module=Storage
time="2024-07-10T14:30:18+02:00" level=trace msg="Query (took 35.917µs): INSERT INTO `did_service` (`id`,`did_document_id`,`data`) VALUES (\"#2\",\"125fedb4-ee25-4933-a28b-55c01c2e435a\",\"{}\") ON CONFLICT (`id`) DO UPDATE SET `did_document_id`=`excluded`.`did_document_id`" module=Storage
time="2024-07-10T14:30:18+02:00" level=trace msg="Query (took 550.208µs): INSERT INTO `did_document_version` (`id`,`did`,`created_at`,`updated_at`,`version`,`raw`) VALUES (\"125fedb4-ee25-4933-a28b-55c01c2e435a\",\"did:web:example.com:iam:bob\",0,1720614618,1,\"{\"\"@context\"\":[\"\"https://w3c-ccg.github.io/lds-jws2020/contexts/lds-jws2020-v1.json\"\",\"\"https://www.w3.org/ns/did/v1\"\"],\"\"assertionMethod\"\":[{\"\"controller\"\":\"\"did:web:example.com:iam:bob\"\",\"\"id\"\":\"\"\"\"}],\"\"authentication\"\":[{\"\"controller\"\":\"\"did:web:example.com:iam:bob\"\",\"\"id\"\":\"\"\"\"}],\"\"capabilityDelegation\"\":[{\"\"controller\"\":\"\"did:web:example.com:iam:bob\"\",\"\"id\"\":\"\"\"\"}],\"\"capabilityInvocation\"\":[{\"\"controller\"\":\"\"did:web:example.com:iam:bob\"\",\"\"id\"\":\"\"\"\"}],\"\"id\"\":\"\"did:web:example.com:iam:bob\"\",\"\"keyAgreement\"\":[{\"\"controller\"\":\"\"did:web:example.com:iam:bob\"\",\"\"id\"\":\"\"\"\"}],\"\"service\"\":[{\"\"id\"\":\"\"\"\"}],\"\"verificationMethod\"\":[{\"\"controller\"\":\"\"did:web:example.com:iam:bob\"\",\"\"id\"\":\"\"\"\"}]}\")" module=Storage
```

Reply:
I compared it to some other foreign key relationships like the one on DIDDocument.DID, but those are "Belongs To" relationships while these are "Has Many" relationships. The direction of relationships are reversed explaining my confusion. This gorm magic naming gets me every time.